onnx/defs/nn/defs.cc

```diff
@@ -2832,4 +2832,470 @@ ONNX_OPERATOR_SET_SCHEMA(
+                return false;
+              int64_t T = tp->tensor_type().elem_type();
+
+              auto type_attr = ctx.getAttribute("stash_type");
+              int64_t U =
+                  (type_attr != nullptr) ? type_attr->i() : static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+              if ((U != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) && (U != ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16))
+                return false; // Error
+
+              auto* axis_attr = ctx.getAttribute("axis");
+              int64_t axis = (axis_attr != nullptr) ? axis_attr->i() : -1;
+              auto* epsilon_attr = ctx.getAttribute("epsilon");
+              float epsilon = (epsilon_attr != nullptr) ? epsilon_attr->f() : 1e-5f;
+
+              auto mktensor = [](int64_t val) -> ONNX_NAMESPACE::TensorProto {
+                auto tp = ONNX_NAMESPACE::ToTensor(std::vector<int64_t>{val});
+                tp.add_dims(1);
+                return tp;
+              };
+              // The treatment of "axis" is different in "RMSNormalization" and in Reduction operations.
+              // This complicates the function definition, requiring reshaping inputs/outputs.
+              // Input X shape: [d[0], ..., d[axis-1], d[axis], ..., d[rank-1]]
+              // This is treated as a 2D shape [d[0] * ... * d[axis-1], d[axis] * ... * d[rank-1]]
+              // Normalization is applied to the second dimension.
+              // Output Y has same shape as X
+              // Outputs InvStdDev have shape: [d[0], ..., d[axis-1], 1, ..., 1]
+              FunctionBuilder builder(functionProto);
+              builder.Const("FloatEpsilon", ToTensor<float>(epsilon))
+                  .Add("Epsilon = Cast (FloatEpsilon)", "to", U)
+                  .Add("XShape = Shape (X)") // shape of input tensor: 1D tensor
+                  .Add("Rank = Size (XShape)") // rank of input tensor: scalar
+                  .Add("Zero1D = Constant()", "value", mktensor(0)) // [0] : 1D tensor
+                  .Add("Axis1D = Constant()", "value", mktensor(axis)) // [axis] : 1D tensor
+                  .Add("PrefixShape = Slice (XShape, Zero1D, Axis1D)") // [d[0], ..., d[axis-1]]
+                  .Add(
+                      axis >= 0 // number of axes that are reduced =
+                          ? "NumReducedAxes = Sub (Rank, Axis1D)" // [rank - axis]: 1D tensor
+                          : "NumReducedAxes = Neg (Axis1D)") // [-axis] : 1D tensor
+                  .Add(
+                      "SuffixShape = ConstantOfShape (NumReducedAxes)",
+                      "value",
+                      mktensor(1)) // [1, ..., 1] for reduced axes
+                  .Add("ReducedShape = Concat <axis = 0> (PrefixShape, SuffixShape)") // [d[0], ..., d[axis-1], 1, ..., 1]
+                  .Add("X2D = Flatten (X)", "axis", axis)
+                  .Add("XU = Cast (X2D)", "to", U);
+              builder.Add("Axes_1 = Constant()", "value", mktensor(1))
```

Comment:
Can we just use `Range(axis,rank)` as the reduction axes, instead of converting to 2D and back? May be worth thinking about pros/cons, which is simpler

Reply:
I think this makes a lot of sense. Updated the logic to avoid the double reshaping that is happening currently